Mathlib/Lean/Meta/RefinedDiscrTree.lean

```diff
@@ -0,0 +1,1204 @@
+  return keys.foldl (insertDTExpr · · v) d
+
+/-- Insert the value `vLhs` at index `lhs`, and if `rhs` is indexed differently, then also
+insert the value `vRhs` at index `rhs`. -/
+def insertEqn [BEq α] (d : RefinedDiscrTree α) (lhs rhs : Expr) (vLhs vRhs : α)
+    (onlySpecific : Bool := true) (config : WhnfCoreConfig := {})
+    (fvarInContext : FVarId → Bool := fun _ => false) : MetaM (RefinedDiscrTree α) := do
+  let keysLhs ← mkDTExprs lhs config onlySpecific fvarInContext
+  let keysRhs ← mkDTExprs rhs config onlySpecific fvarInContext
+  let d := keysLhs.foldl (insertDTExpr · · vLhs) d
+  if @List.beq _ ⟨DTExpr.eqv⟩ keysLhs keysRhs then
+    return d
+  else
+    return keysRhs.foldl (insertDTExpr · · vRhs) d
+
+
+
+/-!
+### Matching with a RefinedDiscrTree
+
+We use a very simple unification algorithm. For all star/metavariable patterns in the
+`RefinedDiscrTree` and in the target, we store the assignment, and when it is assigned again,
+we check that it is the same assignment.
+-/
+
+namespace GetUnify
+
+/-- If `k` is a key in `children`, return the corresponding `Trie α`. Otherwise return `none`. -/
+def findKey (children : Array (Key × Trie α)) (k : Key) : Option (Trie α) :=
+  (·.2) <$> children.binSearch (k, default) (·.1 < ·.1)
+
+private structure Context where
+  unify : Bool
+  config : WhnfCoreConfig
+
+private structure State where
+  /-- Score representing how good the match is. -/
+  score : Nat := 0
+  /-- Metavariable assignments for the `Key.star` patterns in the `RefinedDiscrTree`. -/
+  starAssignments : HashMap Nat DTExpr := {}
+  /-- Metavariable assignments for the `Expr.mvar` in the expression. -/
+  mvarAssignments : HashMap MVarId (Array Key) := {}
+
+
+private abbrev M := ReaderT Context $ StateListM State
+
+/-- Return all values from `x` in an array, together with their scores. -/
+private def M.run (unify : Bool) (config : WhnfCoreConfig) (x : M (Trie α)) :
+    Array (Array α × Nat) :=
+  ((x.run { unify, config }).run {}).toArray.map (fun (t, s) => (t.values!, s.score))
+
+/-- Increment the score by `n`. -/
+private def incrementScore (n : Nat) : M Unit :=
+  modify fun s => { s with score := s.score + n }
+
+/-- Log a metavariable assignment in the `State`. -/
+private def insertStarAssignment (n : Nat) (e : DTExpr) : M Unit :=
+  modify fun s => { s with starAssignments := s.starAssignments.insert n e }
+
+/-- Log a metavariable assignment in the `State`. -/
+private def assignMVar (mvarId : MVarId) (e : Array Key) : M Unit := do
+  let { mvarAssignments, .. } ← get
+  match mvarAssignments.find? mvarId with
+  | some e' => guard (e == e')
+  | none =>
+    modify fun s => { s with mvarAssignments := s.mvarAssignments.insert mvarId e }
+
+/-- Return the possible `Trie α` that match with `n` metavariable. -/
+partial def skipEntries (t : Trie α) (skipped : Array Key) : Nat → M (Array Key × Trie α)
+  | 0      => pure (skipped, t)
+  | skip+1 =>
+    t.children!.foldr (init := failure) fun (k, c) x =>
```

Comment:
Folding with an `Alternative` seems like it should be abstracted, but I guess lean doesn't have `Foldable` types?

Reply:
Ok. I see that `List` has `firstM`. This is essentially `lastM` but for `Array`?

Reply:
I would also prefer if this fold with an alternative could be abstracted. I wasn't aware of `List.firstM`. That is essentially what I would need here, but for `Array`. Having the `Foldable` type class in Lean would solve this :)